tests/components/samsungtv/test_media_player.py

```diff
@@ -152,24 +152,22 @@ def mock_now():
 
 async def test_update_access_denied(hass, remote, mock_now):
     """Testing update tv unhandled response exception."""
+    await setup_samsungtv(hass, MOCK_CONFIG)
```

Comment:
We still need to patch the socket during set up.

Reply:
Maybe I shouldn't do such things not only on github website. :)

Reply:
It is by the remote fixture.

Reply:
I don't get it. It's working on CI but not locally:
```
ERROR    homeassistant.core:core.py:139 Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/home/andreas2/Entwicklung/home-assistant/homeassistant/helpers/storage.py", line 177, in _async_callback_delayed_write
    await self._async_handle_write_data()
  File "/home/andreas2/Entwicklung/home-assistant/homeassistant/helpers/storage.py", line 197, in _async_handle_write_data
    self._write_data, self.path, data
  File "/usr/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "<string>", line 3, in _write_data
  File "/usr/lib/python3.7/unittest/mock.py", line 1011, in __call__
    return _mock_self._mock_call(*args, **kwargs)
  File "/usr/lib/python3.7/unittest/mock.py", line 1077, in _mock_call
    result = effect(*args, **kwargs)
  File "/home/andreas2/Entwicklung/home-assistant/tests/common.py", line 989, in mock_write_data
    data[store.key] = json.loads(json.dumps(data_to_write, cls=store._encoder))
  File "/usr/lib/python3.7/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib/python3.7/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib/python3.7/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/usr/lib/python3.7/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type MagicMock is not JSON serializable
```

Reply:
Found the reason. The ip address is a mock object, this is fixed now.

But now the `CONF_ON_ACTION` is not serializable because it's converted in `__init__` to a script object:
```
INFO     tests.common:common.py:987 Writing data to core.config_entries: {'version': 1, 'key': 'core.config_entries', 'data': {'entries': [{'entry_id': 'ff1a185488d64313be3930a69c436cf8', 'version': 1, 'domain': 'samsungtv', 'title': 'fake', 'data': {'host': 'fake', 'id': None, 'ip_address': 'FAKE_IP_ADDRESS', 'manufacturer': None, 'method': 'websocket', 'model': None, 'name': 'fake', 'turn_on_action': [{'delay': datetime.timedelta(seconds=1)}], 'port': 8001}, 'options': {}, 'system_options': {'disable_new_entities': False}, 'source': 'import', 'connection_class': 'local_poll', 'unique_id': 'FAKE_IP_ADDRESS'}]}}
ERROR    homeassistant.core:core.py:139 Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/home/andreas2/Entwicklung/home-assistant/homeassistant/helpers/storage.py", line 177, in _async_callback_delayed_write
    await self._async_handle_write_data()
  File "/home/andreas2/Entwicklung/home-assistant/homeassistant/helpers/storage.py", line 197, in _async_handle_write_data
    self._write_data, self.path, data
  File "/usr/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "<string>", line 3, in _write_data
  File "/usr/lib/python3.7/unittest/mock.py", line 1011, in __call__
    return _mock_self._mock_call(*args, **kwargs)
  File "/usr/lib/python3.7/unittest/mock.py", line 1077, in _mock_call
    result = effect(*args, **kwargs)
  File "/home/andreas2/Entwicklung/home-assistant/tests/common.py", line 989, in mock_write_data
    data[store.key] = json.loads(json.dumps(data_to_write, cls=store._encoder))
  File "/usr/lib/python3.7/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib/python3.7/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib/python3.7/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/usr/lib/python3.7/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type timedelta is not JSON serializable
```

Reply:
@MartinHjelmare I really don't know how to avoid this. Can we pass the original config before the convert to a script?

Reply:
Hmm, I think the "problem" is already here existing:

https://github.com/home-assistant/home-assistant/blob/a09ac33d3b6c3c3d9809b03ed96dc8390760da46/homeassistant/components/samsungtv/config_flow.py#L117-L120

The script object is assigned to the config_entry and need to be serialized at some point.

Reply:
I don't think we can easily serialize the script config.

It's probably easier to save the script config per host in hass.data in `async_setup` before we load config yaml into the config flow.